File: medium/0665_non_decreasing_array/solution.py
from typing import List


class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        modified = False

        for i in range(len(nums) - 1):
            if nums[i] > nums[i + 1]:
                if modified or (
                    (nums[i - 1] if i != 0 else nums[i + 1]) > nums[i + 1]
                    and nums[i]
                    > (nums[i + 2] if i + 2 < len(nums) else nums[i])
                ):
                    return False
                else:
                    modified = True

        return True

File: medium/0665_non_decreasing_array/solution.cpp
#include <vector>

class Solution {
public:
  bool checkPossibility(std::vector<int> &nums) {
    bool modified = false;

    for (int i = 0; i + 1 < nums.size(); ++i) {
      if (nums[i] > nums[i + 1]) {
        if (modified)
          return false;
        else
          modified = true;

        if (i > 0 && nums[i - 1] > nums[i + 1])
          nums[i + 1] = nums[i];
        else
          nums[i] = nums[i + 1];
      }
    }

    return true;
  }
};
